src/rust/engine/fs/src/glob_matching.rs

```diff
@@ -0,0 +1,397 @@
+        .map(move |path_stats| SingleExpansionResult {
+          sourced_glob,
+          path_stats,
+          globs: vec![],
+        })
+        .to_boxed(),
+      PathGlob::DirWildcard { canonical_dir, symbolic_path, wildcard, remainder } =>
+      // Filter directory listing and request additional PathGlobs for matched Dirs.
+        self.directory_listing(canonical_dir, symbolic_path, wildcard, exclude)
+        .and_then(move |path_stats| {
+          path_stats.into_iter()
+            .filter_map(|ps| match ps {
+              PathStat::Dir { path, stat } =>
+                Some(
+                  PathGlob::parse_globs(stat, path, &remainder)
+                    .map_err(|e| Self::mk_error(e.as_str()))
+                ),
+              PathStat::File { .. } => None,
+            })
+            .collect::<Result<Vec<_>, E>>()
+        })
+        .map(move |path_globs| {
+          let flattened = path_globs
+            .into_iter()
+            .flat_map(|path_globs| path_globs.into_iter())
+            .collect();
+          SingleExpansionResult {
+            sourced_glob,
+            path_stats: vec![],
+            globs: flattened,
+          }
+        })
+        .to_boxed(),
+    }
+  }
+
+  ///
+  /// Canonicalize the Link for the given Path to an underlying File or Dir. May result
+  /// in None if the PathStat represents a broken Link.
+  ///
+  /// Skips ignored paths both before and after expansion.
+  ///
+  /// TODO: Should handle symlink loops (which would exhibit as an infinite loop in expand).
+  ///
+  fn canonicalize(&self, symbolic_path: PathBuf, link: Link) -> BoxFuture<Option<PathStat>, E> {
+    // Read the link, which may result in PathGlob(s) that match 0 or 1 Path.
+    let context = self.clone();
+    self
+      .read_link(link)
+      .map(|dest_path| {
+        // If the link destination can't be parsed as PathGlob(s), it is broken.
+        dest_path
+          .to_str()
+          .and_then(|dest_str| {
+            // Escape any globs in the parsed dest, which should guarantee one output PathGlob.
+            PathGlob::create(&[Pattern::escape(dest_str)]).ok()
+          })
+          .unwrap_or_else(|| vec![])
+      })
+      .and_then(|link_globs| {
+        let new_path_globs =
+          future::result(PathGlobs::from_globs(link_globs)).map_err(|e| Self::mk_error(e.as_str()));
+        new_path_globs.and_then(move |path_globs| context.expand(path_globs))
+      })
+      .map(|mut path_stats| {
+        // Since we've escaped any globs in the parsed path, expect either 0 or 1 destination.
+        path_stats.pop().map(|ps| match ps {
+          PathStat::Dir { stat, .. } => PathStat::dir(symbolic_path, stat),
+          PathStat::File { stat, .. } => PathStat::file(symbolic_path, stat),
+        })
+      })
+      .to_boxed()
+  }
+}
+
+impl<E: Send + Sync + 'static, T: VFS<E>> GlobMatchingImplementation<E> for T {}
```

Comment:
Since both of these have a VFS bound, it's not clear to me that it is worth splitting them (to avoid having abstract methods on GlobMatchingImpl?). IMO: merge them.

Reply:
I would love to -- I was under the impression that keeping e.g. `expand_single()` (and its return type) private would be a goal of part of this. That can still be achieved if e.g. `expand_single()` is e.g. a top-level `fn` -- but if that's not super important, we can just merge them into one. See @illicitonion's comment [here](https://github.com/pantsbuild/pants/pull/5945#discussion_r194366870).

Reply:
Fine with a comment.

Reply:
Done!